Improved the accuracy of delta sum close to the edge of the Fermi window by including energy eigenvalues on-the-fly of the tetrahedral vertices with energies lying outside the chosen Fermi-window and added a test for it

prikarsartam opened the pull request:
Following Issue #99 this commit makes the following **improvements to resolve it across the computation of electronic spectrum using `elphbolt`**. In summary: 

- Added a way to _keep track of the exceptional vertices_ during analytical tetrahedron method.
- Added _a new subroutine to compute the eigenvalues of the tetrahedral vertices on-the-fly_.
- _Updated Fermi-window adjustment across the computation of electronic spectral properties_.
- _Added a test_ to compute the electronic density of states with Fermi-window adjusted analytical tetrahedron method.
(`fpm test fermi_window_adjustment --runner="sh test/3C-SiC/fpm_run_fermi_window_adj_caf.sh"`)
-----

**1. Update in `form_tetrahedra_3d` in `src/delta.f90`** : to keep track of the vertices whose eigenvalues lie outside the chosen Fermi-window by storing their indices as negative.

_Previously_:
```fortran
if(blocks) then
                !Which point in indexlist does aux correspond to?
                call binsearch(indexlist, aux, tmp) !tmp < 0 if search fails.
             end if
             tetra(count, tl) = tmp 
```
is _updated_ to:
```fortran
! If energy-restricted blocks are being used, find the index in indexlist
	    ! The updated snippet keeps track of vertices whose eigenvalues lie outside the chosen window around fermi energy
	    
	    if(blocks) then
		call binsearch(indexlist, aux, tmp)   ! Binary search in indexlist

		
		if (tmp < 0) then
		    tetra(count, tl) = -aux            ! If vertex is outside Fermi window, save negative index (for the exceptional vertices)
		else
		    tetra(count, tl) = tmp             ! Save the index from indexlist
		end if
		
	    else
	    tetra(count, tl) = aux                ! Save the multiplexed index
	    end if
```

--------

**2. Added a new subroutine `fermi_window_adjusted_fill_tetrahedra_3d` in `src/delta.f90`** to compute the eigenvalues of the vertices on-the-fly.

```fortran

subroutine fermi_window_adjusted_fill_tetrahedra_3d(wann, crys, wave_vector_mesh, tetra, evals, tetra_evals)
    type(wannier), intent(in) :: wann
    type(crystal), intent(in) :: crys
    integer(i64), intent(in) :: tetra(:,:)
    real(r64), intent(in) :: evals(:,:)
    integer(i64), intent(in) :: wave_vector_mesh(1, 3)
    
    real(r64), allocatable, intent(out) :: tetra_evals(:,:,:)

    integer(i64) :: iv, it, ib, numbands, aux, numtetra
    integer(i64) :: kpoint(3)
    
    real(r64), allocatable :: energies(:,:)
    real(r64) :: kvecs(1,3)

    ! Calculate dimensions
    numtetra = size(tetra, 1)
    numbands = size(evals, 2)

    print *, "Number of tetrahedra (numtetra) =", numtetra
    print *, "Number of bands (numbands) =", numbands

	  print *, "tetra shape:", shape(tetra)
	  print *, "evals shape:", shape(evals)
	  print *, "tetra_evals shape:", shape(tetra_evals)

    ! Allocate arrays with correct dimensions
   allocate(tetra_evals(numtetra, numbands, 4))
   allocate(energies(1, wann%numwannbands))

    ! Initialize tetra_evals and energies
    tetra_evals(:,:,:) = 0.0_r64
    energies(:, :) = 0.0_r64

    ! Loop over all tetrahedra and vertices
    do it = 1, numtetra
       do iv = 1, 4
          aux = tetra(it, iv)
          
          !print *, "Accessing tetra array with index it=", it, ", iv=", iv

	  !print *, "aux = ", aux

          if (aux < 0) then
             ! Vertex outside Fermi window, calculate on-the-fly
             call demux_vector(-aux, kpoint, wave_vector_mesh, 1_i64)
             kvecs(1, :) = real(kpoint) / real(wave_vector_mesh(1, :))

             call wann%el_wann(crys, 1_i64, kvecs, energies)
             tetra_evals(it, :, iv) = energies(1, :)
          else
             ! Use pre-calculated eigenvalue
             tetra_evals(it, :, iv) = evals(aux, :)
          end if
       end do
    end do

    ! Sort eigenvalues for each tetrahedron vertex
    do it = 1, numtetra
       do ib = 1, numbands
          call sort(tetra_evals(it, ib, :))
       end do
    end do
    deallocate(energies)
end subroutine fermi_window_adjusted_fill_tetrahedra_3d
```

And made it `public`.

--- 
**3. Implemented the fermi-window adjustment in `src/electron.f90`** in computation of electronic spectral properties, by calling `fermi_window_adjusted_fill_tetrahedra_3d` instead of `fill_tetrahedra_3d`; which remains as it is for computing phononic spectral properties.


**NOTE:** that these changes ended up compiling and installing fine, but caused the `electron DOS` test to fail. Then I found out that it compares two computed and reference sets of values within a tolerance of $$10^{-11}$$, so I had printed out the numbers and found them different. So for the new test of this improvement, I took the newly generated data from the `el.dos` file and put them in the same syntax as before inside `test_array(itest)%assert` in my test file `test/check_fermi_window_adjustment.f90`. That way, as expected, it passes.

---  
**4. Finally added the test.**
- Added `test/check_fermi_window_adjustment.f90` similar to `bte_regression.f90`, except with minimum dependencies to test the computation of electronic density of states.
- Added `test/3C-SiC/fpm_run_fermi_window_adj_caf.sh` similar to the `runner` of `bte_regression` test which executes `cafrun -n 2 ../build/caf_*/test/fermi_window_adjustment | tee 3C_SiC_test.output`.
- Updated `fpm.toml` with an addition of 
```toml
[[test]]
name = "fermi_window_adjustment"
source-dir="test"
main = "check_fermi_window_adjustment.f90"
```
---

With these, the fermi-window adjustment can be easily tested by running

`fpm test fermi_window_adjustment --runner="sh test/3C-SiC/fpm_run_fermi_window_adj_caf.sh"`.

Which renders the final output that ends with the following:

```bash
___________________________________________________________________________
___________________________________________Calculating density of states...
Calculating electron density of states...
 electron DOS =>  PASSED! :)
 +--------------------------------------------------+
  Tests carried out: [number of symmetries; symmetry group; electron energies; electron DOS]
  Total number of tests:            4
  Number of tests passed:            4
 +--------------------------------------------------+
..............
| Timing info: elphbolt: Fermi-window adjustment  0.12848908E-03 hr
..............
```
-----

Therefore it Fixes #99

nakib commented on src/delta.f90:
```diff
@@ -586,11 +586,25 @@ subroutine form_tetrahedra_3d(nk, mesh, tetra, tetracount, tetramap, &
+	    
+	    if(blocks) then
+		call binsearch(indexlist, aux, tmp)   ! Binary search in indexlist
+
```
Remove the unnecessary space.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+  
+  
+subroutine fermi_window_adjusted_fill_tetrahedra_3d(wann, crys, wave_vector_mesh, tetra, evals, tetra_evals)
+
```
Remove this space.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+    integer(i64), intent(in) :: tetra(:,:)
+    real(r64), intent(in) :: evals(:,:)
+    integer(i64), intent(in) :: wave_vector_mesh(1, 3)
+    
```
Remove this space.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+
+	  print *, "tetra shape:", shape(tetra)
+	  print *, "evals shape:", shape(evals)
+	  print *, "tetra_evals shape:", shape(tetra_evals)
```
Comment these out.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+
+	  !print *, "aux = ", aux
+
+          if (aux < 0) then
```
In this code base we write `if(condition)`, not `if (condition)`. Remove the space after `if`.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+          if (aux < 0) then
+             ! Vertex outside Fermi window, calculate on-the-fly
+             call demux_vector(-aux, kpoint, wave_vector_mesh, 1_i64)
+             kvecs(1, :) = real(kpoint) / real(wave_vector_mesh(1, :))
```
Casting to real in both the numerator and denominator is an overkill. Doing it just for one is sufficient. Also, please check the documentation of `real` to see if the type of the casting has to be specified. Otherwise, there will be a loss of precision.

nakib replied:
Write `a/b`, not `a / b`. Only use space around and `+`, `-`, not around `*`, `/` or `**`.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+    type(crystal), intent(in) :: crys
+    integer(i64), intent(in) :: tetra(:,:)
+    real(r64), intent(in) :: evals(:,:)
+    integer(i64), intent(in) :: wave_vector_mesh(1, 3)
```
Why are you passing this as a rank-2 tensor? Just `wave_vector_mesh(3)` should be fine. Please check.

nakib replied:
Some names are so obvious from the theoretical context that we don't need such long names. For example, `wvmesh` is fine. For other cases, longer descriptive names are useful.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+   allocate(energies(1, wann%numwannbands))
+
+    ! Initialize tetra_evals and energies
+    tetra_evals(:,:,:) = 0.0_r64
```
It is preferable to to write `tetra_evals(:, :, :)` instead of `tetra_evals(:,:,:)`. This style rule will be strictly enforced in the near future.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+    end do
+    deallocate(energies)
+end subroutine fermi_window_adjusted_fill_tetrahedra_3d
+
```
Remove all of these extra spaces. Just a single space between two procedures will suffice.

nakib commented on src/delta.f90:
```diff
@@ -586,11 +586,25 @@ subroutine form_tetrahedra_3d(nk, mesh, tetra, tetracount, tetramap, &
+	    tetra(count, tl) = aux                ! Save the multiplexed index
+	    end if
+	    
 
```
Remove all of these random extra spaces.

nakib commented on src/delta.f90:
```diff
@@ -586,11 +586,25 @@ subroutine form_tetrahedra_3d(nk, mesh, tetra, tetracount, tetramap, &
+		end if
+		
+	    else
+	    tetra(count, tl) = aux                ! Save the multiplexed index
```
This is not properly indented. If you are using emacs, just select the whole region and press `TAB` to get proper indentation. Other editors should have a similar functionality.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
   end subroutine fill_tetrahedra_3d
+  
+  
+subroutine fermi_window_adjusted_fill_tetrahedra_3d(wann, crys, wave_vector_mesh, tetra, evals, tetra_evals)
```
This name is not clear. Just calling it `fill_tetrahedra_3d_new` is fine. We will replace the old one with this following a demonstration of correctness.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+    real(r64), allocatable, intent(out) :: tetra_evals(:,:,:)
+
+    integer(i64) :: iv, it, ib, numbands, aux, numtetra
+    integer(i64) :: kpoint(3)
```
A better name would be `k_intvec`, since this is an integer triplet.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+    integer(i64) :: kpoint(3)
+    
+    real(r64), allocatable :: energies(:,:)
+    real(r64) :: kvecs(1,3)
```
A better name would be `k_frac`, since this is the wave vector expressed in fractional units.

nakib commented on src/delta.f90:
```diff
@@ -644,6 +658,77 @@ subroutine fill_tetrahedra_3d(tetra, evals, tetra_evals)
+             call demux_vector(-aux, kpoint, wave_vector_mesh, 1_i64)
+             kvecs(1, :) = real(kpoint) / real(wave_vector_mesh(1, :))
+
+             call wann%el_wann(crys, 1_i64, kvecs, energies)
```
This is not a generally applicable solution. You must pass the scissor shift to the `el_wann`.

nakib commented on src/electron.f90:
```diff
@@ -25,7 +25,7 @@ module electron_module
   use crystal_module, only: crystal, calculate_wavevectors_full
   use symmetry_module, only: symmetry, find_irred_wedge, create_fbz2ibz_map
-  use delta, only: form_tetrahedra_3d, fill_tetrahedra_3d, form_triangles, &
+  use delta, only: form_tetrahedra_3d, fermi_window_adjusted_fill_tetrahedra_3d, form_triangles, &
```
Please do not remove the preexisting `fill_tetrahedra_3d`. We are not yet sure if your new implementation is correct.

nakib commented on src/electron.f90:
```diff
@@ -584,7 +584,9 @@ subroutine calculate_electrons(self, wann, crys, sym, num)
        call form_tetrahedra_3d(self%nwv, self%wvmesh, &
             self%simplicial_complex, self%simplex_count, &
             self%simplex_map, .true., self%indexlist)
-       call fill_tetrahedra_3d(self%simplicial_complex, self%ens, self%simplex_evals)
```
Do not remove this line. Just comment it out.

nakib commented on test/check_fermi_window_adjustment.f90:
```diff
@@ -0,0 +1,125 @@
+  use crystal_module, only: crystal
+  use symmetry_module, only: symmetry
+  use electron_module, only: electron
+  use phonon_module, only: phonon
```
This is not needed for your test.

nakib commented on test/check_fermi_window_adjustment.f90:
```diff
@@ -0,0 +1,125 @@
+  use wannier_module, only: wannier
+  use bte_module, only: bte
+  use bz_sums, only: calculate_dos, calculate_qTF, calculate_el_dos_fermi, calculate_el_Ws
+  use interactions, only: calculate_gReq, calculate_gkRp, calculate_3ph_interaction, &
```
This is not needed.

nakib commented on test/check_fermi_window_adjustment.f90:
```diff
@@ -0,0 +1,125 @@
+  use electron_module, only: electron
+  use phonon_module, only: phonon
+  use wannier_module, only: wannier
+  use bte_module, only: bte
```
This is not needed.

nakib commented on test/check_fermi_window_adjustment.f90:
```diff
@@ -0,0 +1,125 @@
+  end if
+  sync all
+  call t_all%end_timer('elphbolt: Fermi-window adjustment')
+end program check_fermi_window_adjustment
```
This program is of limited use and can be removed later. For now, let's keep it as is.

nakib reviewed (changes requested):
Dear Pritam,

Thanks a lot for your contribution. Really appreciate it!

I left my review below. Please take a look when you have time.

prikarsartam commented:
Dear Nakib,

Thank you for all the valuable comments. I have addressed all the suggestions, made the required changes and removed the new test file I added.

Kindly review the latest implementations when you are available.

Thank you.

nakib commented:
Dear Pritam,

Thanks again for your contribution. Really appreciate it.

Best,
Nakib

prikarsartam commented:
Dear Nakib,

I am very happy to see my little contribution merged in `elphbolt`. It will be my pleasure to work on it in the future.

Thank you,
Pritam.
